File: examples/frontend/react/src/services/agent.ts
import { Client } from '@langchain/langgraph-sdk';

export interface MessageContent {
  type: 'text' | 'image_url';
  text?: string;
  image_url?: {
    url: string;
  };
}

export interface AgentMessage {
  role: 'human' | 'ai' | 'tool';
  content: string | MessageContent[];
  type?: string;
  tool_call_id?: string;
  id?: string;
}

export class AgentService {
  private threadId: string;
  private client: Client;
  private assistantId: string;

  constructor(threadId: string, baseUrl?: string) {
    this.threadId = threadId;
    // Convert relative path to full URL for the SDK
    const fullUrl = baseUrl 
      ? (baseUrl.startsWith('http') ? baseUrl : `${window.location.origin}${baseUrl}`)
      : undefined;
    this.client = new Client({ apiUrl: fullUrl });
    this.assistantId = 'ui_agent';
  }

  async maybeCreateThread(): Promise<void> {
    try {
      // Try to get the thread, if it doesn't exist, create it
      try {
        await this.client.threads.get(this.threadId);
      } catch (error: any) {
        if (error.status === 404) {
          await this.client.threads.create({
            threadId: this.threadId,
            ifExists: 'do_nothing',
          });
        } else {
          throw error;
        }
      }
    } catch (error) {
      console.error('Error creating thread:', error);
      throw error;
    }
  }

  async getPastMessages(): Promise<AgentMessage[]> {
    await this.maybeCreateThread();

    try {
      const history = await this.client.threads.getHistory(this.threadId);
      if (history.length > 0) {
        const state = history[0].values as any;
        return state.messages || [];
      }
      return [];
    } catch (error) {
      console.error('Error fetching past messages:', error);
      return [];
    }
  }

  async updateMessage(message: AgentMessage): Promise<void> {
    try {
      const pastMessages = await this.getPastMessages();

      await this.client.threads.updateState(this.threadId, {
        values: {
          messages: [...pastMessages, message],
        },
      });
    } catch (error) {
      console.error('Error updating message:', error);
      throw error;
    }
  }

  async clearHistory(): Promise<void> {
    try {
      await this.client.threads.updateState(this.threadId, {
        values: {
          messages: [],
        },
      });
    } catch (error) {
      console.error('Error clearing history:', error);
      throw error;
    }
  }

  private async fileToBase64(file: File): Promise<string> {
    return new Promise((resolve, reject) => {
      const reader = new FileReader();
      reader.onload = () => {
        const result = reader.result as string;
        // Extract base64 part (remove data:image/...;base64, prefix)
        const base64 = result.split(',')[1];
        resolve(base64);
      };
      reader.onerror = reject;
      reader.readAsDataURL(file);
    });
  }

  async getResponse(
    userInput: string | { text: string; files: File[] }
  ): Promise<AgentMessage[]> {
    try {
      const pastMessages = await this.getPastMessages();

      let messageContent: string | MessageContent[];

      if (typeof userInput === 'string') {
        messageContent = userInput;
      } else {
        // Convert to multi-content format
        const content: MessageContent[] = [];

        if (userInput.text) {
          content.push({
            type: 'text',
            text: userInput.text,
          });
        }

        // Convert files to base64 and add to content
        for (const file of userInput.files) {
          if (file.type.startsWith('image/')) {
            const base64 = await this.fileToBase64(file);
            content.push({
              type: 'image_url',
              image_url: {
                url: `data:${file.type};base64,${base64}`,
              },
            });
          }
        }

        messageContent = content;
      }

      const runResult = await this.client.runs.wait(
        this.threadId,
        this.assistantId,
        {
          input: {
            messages: [
              {
                role: 'human',
                content: messageContent,
              },
            ],
          },
        }
      );

      const runMessages = (runResult as any).messages as AgentMessage[];
      return runMessages.slice(pastMessages.length);
    } catch (error) {
      console.error('Error getting response:', error);
      throw error;
    }
  }

  async *getStreamingResponse(
    userInput: string | { text: string; files: File[] }
  ): AsyncGenerator<AgentMessage, void, unknown> {
    try {
      let messageContent: string | MessageContent[];

      if (typeof userInput === 'string') {
        messageContent = userInput;
      } else {
        // Convert to multi-content format
        const content: MessageContent[] = [];

        if (userInput.text) {
          content.push({
            type: 'text',
            text: userInput.text,
          });
        }

        // Convert files to base64 and add to content
        for (const file of userInput.files) {
          if (file.type.startsWith('image/')) {
            const base64 = await this.fileToBase64(file);
            content.push({
              type: 'image_url',
              image_url: {
                url: `data:${file.type};base64,${base64}`,
              },
            });
          }
        }

        messageContent = content;
      }

      const streamResponse = this.client.runs.stream(
        this.threadId,
        this.assistantId,
        {
          input: {
            messages: [
              {
                role: 'human',
                content: messageContent,
              },
            ],
          },
        }
      );

      for await (const chunk of streamResponse) {
        if (chunk.event === 'messages/partial' && chunk.data) {
          const messages = chunk.data as AgentMessage[];
          for (const message of messages) {
            yield message;
          }
        }
      }
    } catch (error) {
      console.error('Error getting streaming response:', error);
      throw error;
    }
  }
}

File: examples/frontend/streamlit/agent.py
"""Agent for the Streamlit demo."""

import json
import uuid
from pprint import pprint
from langchain_core.messages import HumanMessage
from langgraph_sdk import get_client


class Agent:
    """Agent wrapper."""

    def __init__(self, thread_id: str) -> None:
        """Initialize the agent."""
        self.thread_id = thread_id
        self.client = get_client(url="http://localhost:2024")

    async def maybe_create_thread(self) -> None:
        """Create a thread if it doesn't exist."""
        try:
            await self.client.threads.get(self.thread_id)
        except Exception:
            await self.client.threads.create(
                thread_id=self.thread_id, if_exists="do_nothing"
            )

    async def get_past_messages(self) -> list[dict]:
        """Get past messages from an existing agent."""
        await self.maybe_create_thread()
        past_state = await self.client.threads.get_history(self.thread_id)
        if len(past_state) > 0:
            return past_state[0]["values"]["messages"]
        return []

    async def update_message(self, message: dict) -> None:
        """Update a message.

        Old message will be replaced according to the message id.
        If the message id is not found, the message will be appended to the end
        of the messages.

        Args:
            message: The message to update.
        """
        past_messages = await self.get_past_messages()
        await self.client.threads.update_state(
            self.thread_id,
            values={"messages": past_messages + [message]},
        )

    async def get_response(
        self,
        user_input: HumanMessage,
    ) -> list[dict]:
        """Get response from an existing agent and return new messages."""
        past_messages = await self.get_past_messages()
        response = await self.client.runs.wait(
            self.thread_id,
            "ui_agent",  # Name of assistant. Defined in langgraph.json.
            input={
                "messages": [
                    user_input.model_dump(),
                ],
            },
        )
        return response["messages"][len(past_messages) :]


if __name__ == "__main__":
    import asyncio
    from dotenv import load_dotenv

    load_dotenv()

    def update_tool_data(message: dict) -> dict:
        """Update the tool data."""
        data = json.loads(message["content"])
        data["value"] = 10
        message["content"] = json.dumps(data)
        return message

    async def main():
        """Run the demo agent."""
        agent = Agent(thread_id=str(uuid.uuid4()))

        result = await agent.get_response(
            "Generate a number input between 0 and 100",
        )
        pprint(result)

        updated_tool_message = None
        for msg in result:
            if msg["type"] == "tool":
                updated_tool_message = update_tool_data(msg)
                await agent.update_message(updated_tool_message)

        result = await agent.get_response(
            "What is the value of the number input?",
        )
        pprint(result)

        # result = await agent.get_response(
        #     "What did I just ask?",
        # )
        # pprint(result)

        # result = await agent.get_response(
        #     "What is the value of the number input?",
        # )
        # pprint(result)

    asyncio.run(main())
